### DISCUSSION TITLE
GERS ID encoding and general questions

### DISCUSSION THREAD

**Discussion by fklebert** [OvertureMaps/data#299] [Discussion]
I haven't found anything on the encoding of the planned GERS ID.
Is this going to be something like a UUID (RFC 4122 version 5) or a semantic identifier?

Will there also be a predefined reference to it which for example includes validity and direction on road segments, or is this something an external e.g. traffic provider has to come up with themselves?

I wonder if the GERS ID of a long road segment will be useful. Or is it planned to have multiple GERS IDs per road segment? Just in case a road is e.g. a couple of kilometers long. How would I be able match data to the first 200 meters only.

**Comment by missinglink** [comment]
I'm curious about lifecycle management of GERS IDs, such as when an ID is deprecated, superseded, split, joined, etc.

eg. cases where the change is 'significant enough' for the ID to now be considered a 'different thing' than it previously was.
eg. providing a reference to a new GERS ID which superseded a prior ID, so systems can be updated (ie. `HTTP 301`)
eg. merging of duplicates, etc.

**Reply by jenningsanderson** [reply]
Thanks for the interest, and great questions: 
_Short answer:_ We're still exploring some very different directions, hence not yet proposing the actual ID structure nor a better description of lifecycle management.
_Long answer:_ These are the different directions and open questions around GERS:

#### ID Stability & Traceability
A core tenet of GERS is ID stability. In cases where change is _significant enough_ to warrant a new ID, then GERS needs to be traceable. "Strive for stability, but at the very least, be traceable when stability is impossible"
There will be many cases for traceability: merging of duplicates, higher resolution data (1 building is actually 4 separate buildings, etc.). 

Open Questions: What is the best way to store / share / reference the traceability of GERS? Is the "history" of an entity encoded in the entity itself? Is there a lookup service? Is GERS something that is published at a specific cadence with references to any changes from version n-1? 
Furthermore, what _is_ encoded in GERS? Is it simply a mapping of IDs between datasets, such as: 
```json
"GERS:building:123" : {"OSM_building":456, "MS_building":789, "ESRI building":1011}
```
Or, does it hold slightly more information:
```json
"GERS:building:123" : { 
  "near" : [-130,43],
  "references" : {"OSM_building":456, "MS_building":789, "ESRI building":1011}
}
``` 
Which begs the question, can any of this additional information actually be encoded into the ID itself? 

#### GERS ID Structure
Currently we're exploring a 128bit GERS ID that would encode an entity's [H3 cell](https://www.uber.com/en-FR/blog/h3/) in the first 64 bits with the remaining 64 bits to describe the `theme`, `type`, a unique identifier, and maybe additional per-theme attributes as relevant. For example, should a GERS ID for a [transportation connector](https://docs.overturemaps.org/reference/transportation/connector) look different from the GERS ID for a [transportation segment](https://docs.overturemaps.org/reference/transportation/segment)?

As for OpenLR (mentioned below), yes, this is a great path forward to referencing specific segments within entities.

We really appreciate any feedback you have on these topics!

**Reply by fklebert** [reply]
Encoding a cell or tile identifier into a global and stable identifier somehow sounds contradicting. Think of a business POI which was located on the edge of a cell/tile and it simply moves down the street a bit into another cell/tile. Would it get a new ID? If so, what's the benefit of such an ID in the first place then? The entity itself did not change...
In case of hierarchical cells (like H3) or tiles: Which level/resolution is encoded in GERS? In case you go for most detailed, what about large entities that span across multiple cells?

**Reply by varapmsft** [reply]
I understand why it could be confusing, but the intention of including cell identifier to GERS ID is for readability purposes and we do not intend to be super strict about keeping location of an entity and the cell consistent all the time, in rare cases they may not match. One of the core tenets of GERS is ID stability so the GERS ID will NOT change if the location is corrected/changed and  the entity falls into a new cell.

**Reply by missinglink** [reply]
Could you please elaborate on the meaning of 'readability purposes', I'm trying to understand the trade-off you're proposing and I don't yet understand the value of this readability aspect?

**Reply by varapmsft** [reply]
Sure. By readability we mean the IDs of entities that are located close would look similar instead of just pure random numbers. For example ids of two buildings located next to each other would look very similar. Similarly ids of a building and the road it is located next to would look similar. However I agree it is debatable how important this aspect is, especially if it causes consumers to use it for unintended purposes. As Jennings mentioned above we are still deliberating internally on finalizing the ID structure.

**Comment by msbarry** [comment]
It will be much easier for downstream applications (like planetiler) if element IDs are numeric. One thing about the openstreetmap data model that makes it efficient to process is that all nodes, ways, and relations have 8-byte IDs and because they are sequentially and less than 20 billion they can be stored in less than 8 bytes.

Ideally it seems like overture would provide their own sequentially-increasing IDs for anything that has been incorporated into overture since overture is aware of all of the elements in the corpus, but let the elements have string external IDs that relate to other datasets like wikipedia, openstreetmap, or the original ID used by the provider of the data.

**Reply by fklebert** [reply]
Sorry, just deleted my comment after a second look. You are of course right that 64 bits are enough to address all of those.
Still a UUID would give some benefits in case you want to ingest data in a non-sequential way, because it is more or less collision free.

**Comment by hrabbey** [comment]
> How would I be able match data to the first 200 meters only.

[Geometric Scoping: Linear Referencing](https://docs.overturemaps.org/themes/transportation/#geometric-scoping-linear-referencing) would provide this functionality. [OpenLR](http://www.openlr.org/method.html) is a similar example which has been used in-market for distributing traffic data.

**Comment by ringsaturn** [comment]
I used to learn a lot from another project, named [PlaceKey](https://www.placekey.io), which uses a similar solution.

But only the geo location part has public docs to explain, which based on Uber H3, but encoded to more readable text.
It’s white paper is [here](https://uploads-ssl.webflow.com/5f08ccbb93b299154d34ef7f/6010f1245d848a2c0a40f02f_Placekey%20Technical%20White%20Paper.pdf).

<img width="1160" alt="image" src="https://github.com/OvertureMaps/schema/assets/13536789/a88e3f83-aae4-4013-85c2-23d0756fa3bb">


As discussuib [above](https://github.com/OvertureMaps/schema/discussions/10#discussioncomment-6265447), I hope there could be a public defination for how to build 'remaining 64 bits', since PlaceKey doesn't explain that.

**Reply by auren** [reply]
@ringsaturn good call.  I helped develop Placekey and currently working on a Placekey 2.0 (which will be global and much more scalable).  Placekey 2.0 will launch in a month.  would be great to combine efforts with GERS.
